_posts/contests/2021-02-05-Performance-analysis-and-tuning-contest-4.md

```diff
@@ -96,4 +96,188 @@ __P.S.__ I'm also open to your comments and suggestions. Especially if you have
 
 ---
 
+## _Updated TODO(Add date)_
+
+TODO(Write an introduction about the submission. Put the link to the youtube channel)
+
+### Observations
+
+The profiler showed four functions dominating the profile: `gaussian_smooth`, `derivative_x_y`, `non_max_supp` and `apply_hysteresis`. Let's dig into the function, see what they are doing and what our contestants did to make them faster.
+
+**gaussian_smooth**
+
+The function `gaussian_smooth` (TODO: Add code to the baseline in the repository) consists of two loop nests. Here is the source code of the first loop nest:
+
+```cpp
+for(r=0;r<rows;r++){
+  for(c=0;c<cols;c++){
+      dot = 0.0;
+      sum = 0.0;
+      for(cc=(-center);cc<=center;cc++){
+        if(((c+cc) >= 0) && ((c+cc) < cols)){
+            dot += (float)image[r*cols+(c+cc)] * kernel[center+cc];
+            sum += kernel[center+cc];
+        }
+      }
+      tempim[r*cols+c] = dot/sum;
+  }
+}
+```
+
+The access pattern for the arrays in the innermost loop is good, it is always the sequential pattern. Everytime the loop counter `cc` in the innermost loop increases by one, the index into the `image` and `kernel` increases by one.
+
+The problem is the trip count of the innermost loop: it is low. Vectorization doesn't pay off in that case.
+
+One of the proposed solution was to multiversion a loop. In our case, `center` had value 2. We can create two versions of the loop nest, one where the value of `center` is known at compile time, and another one where it isn't. The compilers are very efficient at unrolling the loop with compile-time known trip count. Here is the solution:
```

Comment:
> In our case, `center` had value 2

I suggest adding how did you find that. And write that we multiversion just because we know the runtime value of `center`. Such specialization doesn't work every time but it helps in our case.